src/providers/openstack/configdrive.rs

```diff
@@ -0,0 +1,166 @@
+//! configdrive metadata fetcher for OpenStack
+//! reference: https://docs.openstack.org/nova/latest/user/metadata.html
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+
+use openssh_keys::PublicKey;
+use slog_scope::{error, warn};
+use tempfile::TempDir;
+
+use crate::errors::*;
+use crate::network;
+use crate::providers::MetadataProvider;
+
+const CONFIG_DRIVE_LABEL: &str = "config-2";
+
+/// OpenStack config-drive.
+#[derive(Debug)]
+pub struct OpenstackConfigDrive {
+    /// Path to the top directory of the mounted config-drive.
+    drive_path: PathBuf,
+    /// Temporary directory for own mountpoint (if any).
+    temp_dir: Option<TempDir>,
+}
+
+impl OpenstackConfigDrive {
+    /// Try to build a new provider client.
+    ///
+    /// This internally tries to mount (and own) the config-drive.
+    pub fn try_new() -> Result<Self> {
+        // Short-circuit if the config-drive is already mounted.
+        let path = Path::new("/media/ConfigDrive/openstack/latest/");
+        if path.exists() {
+            return Ok(OpenstackConfigDrive {
+                temp_dir: None,
+                drive_path: PathBuf::from("/media/ConfigDrive/"),
+            });
+        }
+
+        // Otherwise, try and mount with the label.
+        let target = tempfile::Builder::new()
+            .prefix("afterburn-")
+            .tempdir()
+            .chain_err(|| "failed to create temporary directory")?;
+        crate::util::mount_ro(
+            &Path::new("/dev/disk/by-label/").join(CONFIG_DRIVE_LABEL),
+            target.path(),
+            "iso9660",
+            3,
+        )?;
+
+        let cd = OpenstackConfigDrive {
+            drive_path: target.path().to_owned(),
+            temp_dir: Some(target),
+        };
+        Ok(cd)
+    }
+
+    /// Return the path to the metadata directory.
+    fn metadata_dir(&self, platform: &str) -> PathBuf {
+        self.drive_path.clone().join(platform).join("latest")
+    }
+
+    // The metadata is stored as key:value pair in ec2/meta-data.json file
+    fn fetch_value(&self, key: &str) -> Result<Option<String>> {
+        let filename = self.metadata_dir("ec2").join("meta-data.json");
```

Comment:
It would be good to have a real-world `meta-data.json` fixture and a test to cover this (see `ibmcloud-classic` for an existing example).

Reply:
Added `tests/fixtures/openstack-config-drive/ec2/meta_data.json`.